File: app/src/main/java/org/skyfaced/mvp/ui/MainPresenter.java
package org.skyfaced.mvp.ui;

import org.skyfaced.mvp.model.ImageDto;
import org.skyfaced.mvp.mvp.own.BasePresenter;
import org.skyfaced.mvp.service.instant.InstantService;
import org.skyfaced.mvp.service.network.WaifuRepository;
import org.skyfaced.mvp.util.WaifuType;

import java.util.Collections;
import java.util.List;
import java.util.Random;

import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers;
import io.reactivex.rxjava3.annotations.NonNull;
import io.reactivex.rxjava3.core.Observer;
import io.reactivex.rxjava3.disposables.Disposable;
import io.reactivex.rxjava3.schedulers.Schedulers;
import timber.log.Timber;

public final class MainPresenter extends BasePresenter<MainView> {
    private final WaifuRepository waifuRepository;
    private final InstantService instantService;

    public MainPresenter(WaifuRepository waifuRepository, InstantService instantService) {
        this.waifuRepository = waifuRepository;
        this.instantService = instantService;
    }

    void changeText(String text) {
        Timber.d(text);
    }

    void onItemClick(String item) {
        getView().showSnackbar("Hello from presenter\n" + item);
    }

    void onMessageClick() {
        Integer random = new Random().nextInt();
        if (random % 2 == 0) {
            getView().showSnackbar(random + " from snack");
        } else {
            getView().showToast(random + " from toast");
        }
    }

    List<String> strings() {
        List<String> items = instantService.listOfStrings();
        Collections.shuffle(items);
        return items;
    }

    void onWaifuClick(WaifuType type) {
        getView().showOnScreenLoader();
        waifuRepository.waifu(type)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(
                        new Observer<ImageDto>() {
                            @Override
                            public void onSubscribe(@NonNull Disposable d) {
                            }

                            @Override
                            public void onNext(@NonNull ImageDto image) {
                                getView().updateWaifu(image);
                            }

                            @Override
                            public void onError(@NonNull Throwable e) {
                                getView().showSnackbar(e.getMessage());
                                getView().hideOnScreenLoader();
                            }

                            @Override
                            public void onComplete() {
                                getView().showToast("Вайфу загружена");
                                getView().hideOnScreenLoader();
                            }
                        }
                );
    }

    void onWaifuItemClick(ImageDto image) {
        getView().showSnackbar(image.getUrl());
    }

    void onWaifusClick(WaifuType type) {
        waifuRepository.waifus(type)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Observer<List<ImageDto>>() {
                    @Override
                    public void onSubscribe(@NonNull Disposable d) {

                    }

                    @Override
                    public void onNext(@NonNull List<ImageDto> images) {
                        getView().updateWaifuRecycler(images);
                    }

                    @Override
                    public void onError(@NonNull Throwable e) {
                        getView().showSnackbar(e.getMessage());
                    }

                    @Override
                    public void onComplete() {
                        getView().showToast("Вайфу загружены");
                    }
                });
    }
}

File: app/src/main/java/org/skyfaced/mvp/ui/MainPresenter.kt
package org.skyfaced.mvp.ui

import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers
import io.reactivex.rxjava3.disposables.CompositeDisposable
import io.reactivex.rxjava3.disposables.Disposable
import io.reactivex.rxjava3.schedulers.Schedulers
import moxy.InjectViewState
import moxy.MvpPresenter
import org.skyfaced.mvp.model.ImageDto
import org.skyfaced.mvp.service.instant.InstantService
import org.skyfaced.mvp.service.network.WaifuRepository
import org.skyfaced.mvp.util.WaifuType
import timber.log.Timber
import java.util.*
import java.util.concurrent.TimeUnit
import javax.inject.Inject

@InjectViewState
class MainPresenter @Inject constructor(
    private val waifuRepository: WaifuRepository,
    private val instantService: InstantService,
) : MvpPresenter<MainView>() {
    private val disposable = CompositeDisposable()

    fun changeText(text: String?) {
        Timber.d(text)
    }

    fun onItemClick(item: String) {
        viewState.showSnackbar("Hello from presenter\n$item")
    }

    fun onMessageClick() {
        val random = Random().nextInt()
        if (random % 2 == 0) {
            viewState.showSnackbar("$random from snack")
        } else {
            viewState.showToast("$random from toast")
        }
    }

    fun strings(): List<String> = instantService.strings.shuffled()

    fun onWaifuClick(type: WaifuType) {
        val d: Disposable = waifuRepository.waifu(type)
            .debounce(300, TimeUnit.MILLISECONDS)
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe(
                { image -> viewState.updateWaifu(image) },
                { cause -> viewState.showSnackbar(cause.message.orEmpty()) }
            ) { viewState.showToast("Вайфу загружена") }
        disposable.add(d)
    }

    fun onWaifuItemClick(image: ImageDto) {
        viewState.showSnackbar(image.url)
    }

    fun onWaifusClick(type: WaifuType) {
        viewState.showOnScreenLoader()
        val d: Disposable = waifuRepository.waifus(type)
            .debounce(300, TimeUnit.MILLISECONDS)
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe(
                { images -> viewState.updateWaifuRecycler(images) },
                { cause ->
                    viewState.showSnackbar(cause.message.orEmpty())
                    viewState.hideOnScreenLoader()
                }
            ) {
                viewState.showToast("Вайфу загружены")
                viewState.hideOnScreenLoader()
            }
        disposable.add(d)
    }

    override fun onDestroy() {
        disposable.clear()
    }
}